Added perfect number check algorithm

hefty-head opened the pull request:
Wrote a simple program that checks whether or not a number is perfect

hefty-head committed:
Added perfect number check algorithm

Wrote a simple program that checks whether or not a number is perfect

faheel commented on NumberTheory/PerfectNumbers.cpp:
```diff
@@ -0,0 +1,35 @@
+/*
+    Perfect number:
+    Checks if a given number is perfect. Could be used to generate a 
+    sequence of perfect numbers
```
It would be good if you could add a 1-2 line description of what a perfect number is.

faheel commented on NumberTheory/PerfectNumbers.cpp:
```diff
@@ -0,0 +1,35 @@
+	{
+		if(num % divisor == 0)
+			sum += divisor;
+	}
```
The divisors can be found in _O_(sqrt(N)) by checking only upto sqrt(N), and adding `divisor` and `num / divisor` to `sum`.

faheel replied:
You have changed the complexity but have forgotten to update the code here.

faheel commented on NumberTheory/PerfectNumbers.cpp:
```diff
@@ -0,0 +1,35 @@
+
+using namespace std;
+
+ULL is_perfect(int num) {
```
The return type should be `bool`.

faheel reviewed (changes requested):
It would be nice if can modify the program to print all perfect numbers upto a given number.

hefty-head reviewed (commented):
Will do, thanks

hefty-head committed:
Update PerfectNumbers.cpp

hefty-head committed:
Updated code with edits

hefty-head commented:
Just updated the code, thanks for the notes

hefty-head committed:
Updated Perfect Numbers.cpp with changes

Now prints n perfect numbers, Added description of what a perfect number is. Reduced to O(sqrt(N)) time.

hefty-head reviewed (commented):
I accidentally committed the wrong code whoops

faheel commented on NumberTheory/PerfectNumbers.cpp:
```diff
@@ -0,0 +1,37 @@
+
+bool is_perfect(int num) {
+	int sum = 1;
+	for(int divisor = 2; divisor <= num/2; divisor++)
```
`num`, `sum` and `divisor` should be of type `unsigned long long` so that they can go upto 2<sup>64</sup>-1.

faheel commented on NumberTheory/PerfectNumbers.cpp:
```diff
@@ -0,0 +1,37 @@
+int main() {
+	int num;
+	cout << "The number of perfect numbers you want printed: ";
+	cin >> num;
```
Instead of asking for
> the number of perfect numbers to be printed

ask for
> the number upto which perfect numbers are to be found

as there are only 8 perfect numbers between 1 and 2<sup>64</sup>-1.

You would also need to change `num` to `limit` or something more meaningful, and also it's type to `unsigned long long` to support a higher range of values.

faheel commented on NumberTheory/PerfectNumbers.cpp:
```diff
@@ -0,0 +1,37 @@
+	cout << "The number of perfect numbers you want printed: ";
+	cin >> num;
+	for(int count = 1; count < num; count++)
+	//start from one to prevent false positive with 0
```
This loop would also need to be modified accordingly.

eskeype commented:
Hi @faheel and @aaboaf, I was looking at this commit, and i noticed that the complexity of displaying all of the perfect numbers less than N is O(sqrt(N) * N). However, it we were to use a sieve method this could be done in O(N * log(N)) time. 

In particular, we could have a large int array (to eventually contain the sum of the divisors of each index) with index values up to N, and for each index i>=0 we could loop through all of the multiples of i in the array and add i to their value.

This sieve has space complexity of O(N), which might make it a less attractive option, as there are very few perfect numbers in the very large range of [1..(2^64) - 1]. However, this algorithm could be adapted to work as an efficient method to determine whether numbers are deficient or abundant as well.

faheel commented:
@eskeype Your method would definitely be more efficient in finding perfect numbers but at a very high cost of memory space. To find out the perfect numbers between 1 and 2<sup>27</sup> only, we would require about 1 GB of memory. That's a lot of memory for finding just 5 perfect numbers!

faheel commented:
@aaboaf I now realise that keeping the program as simple as it originally was would be the best, because finding all perfect numbers within a large range would either require a lot of memory if done very efficiently, or a lot a processing time if done not so efficiently.
So just determining whether a given number is perfect or not in _O_(√N) time and _O_(1) space is feasible even for very large values of N (upto 2<sup>64</sup>-1).

hefty-head commented:
Will do!

hefty-head committed:
Updated perfectNumbers.cpp

Restored to original but with sqrt(n) time check and with a larger range of test values

faheel commented on NumberTheory/PerfectNumbers.cpp:
```diff
@@ -0,0 +1,38 @@
+ 	for(divisor = 2; divisor <= sqrt(num); divisor++)
+ 	{
+ 		if(num % divisor == 0)
+ 			sum += divisor;
```
Should be
```C++
sum += divisor + num / divisor;
```

faheel commented on NumberTheory/PerfectNumbers.cpp:
```diff
@@ -0,0 +1,38 @@
+ 	}
+ 	if(divisor == sum)
+ 		return 1;
+ 	return 0;
```
Return `true` and `false` instead of `1` and `0`.

faheel commented on NumberTheory/PerfectNumbers.cpp:
```diff
@@ -0,0 +1,38 @@
+ 		if(num % divisor == 0)
+ 			sum += divisor;
+ 	}
+ 	if(divisor == sum)
```
The condition should be 
```C++
if (sum == num)
```

faheel commented on NumberTheory/PerfectNumbers.cpp:
```diff
@@ -0,0 +1,38 @@
+ 
+ int main() {
+ 	ULL num;
+ 	cout << "Enter a number: ";
```
Ask the user to `"Enter a positive integer less than 2^64 : "` to be precise.

faheel commented on NumberTheory/PerfectNumbers.cpp:
```diff
@@ -0,0 +1,38 @@
+ 	if(is_perfect(num))
+ 		cout << num << " is perfect" << endl;
+ 	else 
+ 		cout << num << " is not perfect" << endl;
```
"is/is not **a** perfect **number**"

faheel commented on NumberTheory/PerfectNumbers.cpp:
```diff
@@ -0,0 +1,38 @@
+ 
+typedef unsigned long long int ULL;
+
+ ULL is_perfect(ULL num) {
```
Return type should be `bool`.

faheel replied:
@aaboaf You missed this :smile:

hefty-head committed:
Update

hefty-head committed:
Bool

I miss C. I miss booleans not being their own thing

faheel commented:
Thank you for your time and contribution @aaboaf! And congratulations on your first PR merge :smiley:
